src/stream/src/executor/nested_loop_join.rs

```diff
@@ -0,0 +1,516 @@
+                    #[for_await]
+                    for chunk in Self::eq_join_oneside::<{ SideType::Left }>(JoinArgs {
+                        ctx: &self.ctx,
+                        side_l: &mut self.side_l,
+                        side_r: &mut self.side_r,
+                        actual_output_data_types: &self.actual_output_data_types,
+                        cond: &mut self.cond,
+                        chunk,
+                        chunk_size: self.chunk_size,
+                    }) {
+                        left_time += left_start_time.elapsed();
+                        yield Message::Chunk(chunk?);
+                        left_start_time = Instant::now();
+                    }
+                    left_time += left_start_time.elapsed();
+                    left_join_match_duration_ns.inc_by(left_time.as_nanos() as u64);
+                    self.try_flush_data().await?;
+                }
+                AlignedMessage::Right(chunk) => {
+                    let mut right_time = Duration::from_nanos(0);
+                    let mut right_start_time = Instant::now();
+                    #[for_await]
+                    for chunk in Self::eq_join_oneside::<{ SideType::Right }>(JoinArgs {
+                        ctx: &self.ctx,
+                        side_l: &mut self.side_l,
+                        side_r: &mut self.side_r,
+                        actual_output_data_types: &self.actual_output_data_types,
+                        cond: &mut self.cond,
+                        chunk,
+                        chunk_size: self.chunk_size,
+                    }) {
+                        right_time += right_start_time.elapsed();
+                        yield Message::Chunk(chunk?);
+                        right_start_time = Instant::now();
+                    }
+                    right_time += right_start_time.elapsed();
+                    right_join_match_duration_ns.inc_by(right_time.as_nanos() as u64);
+                    self.try_flush_data().await?;
+                }
+                AlignedMessage::Barrier(barrier) => {
+                    let barrier_start_time = Instant::now();
+                    // Update the vnode bitmap for state tables of both sides if asked.
+                    let vnode_bitmap = barrier.as_update_vnode_bitmap(self.ctx.id);
+                    let (side_l_post_commit, _) = self.flush_data(barrier.epoch).await?;
+                    if let Some((_, cache_may_stale)) = side_l_post_commit
+                        .post_yield_barrier(vnode_bitmap.clone())
+                        .await?
+                        && cache_may_stale
+                    {
+                        self.watermark_buffers
+                            .values_mut()
+                            .for_each(|buffers| buffers.clear());
+                    }
+                    // Do not update vnode bitmap for right side because it is a broadcast side.
```

Comment:
Not sure it should work this way. I guess we didn't do this for any other executor before>

Reply:
Can we support a singleton distribution first? I think it is identical to the hack of adding dummy columns to the both side of hash join.